CRM/ManualDirectDebit/Page/BatchSubmissionQueue.php

```diff
@@ -0,0 +1,169 @@
+<?php
+
+class CRM_ManualDirectDebit_Page_BatchSubmissionQueue extends CRM_Core_Page {
+
+  private $batchRecordsLimit;
+
+  private $queue;
+
+  private $batchId;
+
+  private $taskItemRecords = [];
+
+  public function __construct($title = NULL, $mode = NULL) {
+    parent::__construct($title, $mode);
+
+    $this->queue = CRM_ManualDirectDebit_Queue_BatchSubmission::getQueue();
+    $this->batchId =  CRM_Utils_Request::retrieveValue('batchId', 'Int');
+    $this->batchRecordsLimit = CRM_ManualDirectDebit_Common_SettingsManager::getBatchSubmissionRecordsPerTaskLimit();
+  }
+
+  public function run() {
+    if (!$this->validBatchStatus()) {
+      CRM_Core_Session::setStatus('The batch is not in a valid status for submission', '', 'error');
+
+      return;
+    }
+
+    $this->addTasksToQueue();
+    $this->runQueue();
+  }
+
+  private function validBatchStatus() {
+    return in_array($this->getBatchStatus(), ['Open', 'Reopened']);
+  }
+
+  private function getBatchStatus() {
+    $statusID = CRM_Core_DAO::getFieldValue('CRM_Batch_DAO_Batch', $this->batchId, 'status_id');
+    $batchStatuses = CRM_Core_PseudoConstant::get('CRM_Batch_DAO_Batch', 'status_id', [
+      'labelColumn' => 'name',
+      'status' => " v.value={$statusID}",
+    ]);
+    return $batchStatuses[$statusID];
+  }
+
+  private function addTasksToQueue() {
+    $batchType = $this->getBatchType();
+    switch ($batchType) {
+      case 'instructions_batch':
+        $this->addInstructionsQueueTasks();
+        break;
+      case 'dd_payments':
+        $this->addPaymentsQueueTasks();
+        break;
+    }
+
+    $this->addBatchSubmissionCompletionTask();
+  }
+
+  private function getBatchType() {
+    $batch = CRM_Batch_DAO_Batch::findById($this->batchId);
+    $batchTypes = CRM_Core_PseudoConstant::get('CRM_Batch_DAO_Batch', 'type_id', ['labelColumn' => 'name']);
+    return $batchTypes[$batch->type_id];
+  }
+
+  private function addInstructionsQueueTasks() {
```

Comment:
Do you think it might be a good idea to refactor these into separate classes? Seems this one is trying to do two thngs at the same time.  Somthing like:

```php
private function addInstructionsQueueTasks() {
  $queueBuilder = new InstructionsQueueBuilder();
  $queueBuilder->buildQueue();
}

private function addPaymentsQueueTasks() {
  $queueBuilder = new PaymentsQueueBuilder();
  $queueBuilder->buildQueue();
}
```

You could even refactor similar code into an abstract class. For example, `addInstructionsQueueTaskItem()` and `addPaymentQueueTaskItem()` are practically identical.

Reply:
yeah! good point